axlearn/audio/asr_decoder.py

```diff
@@ -28,6 +28,40 @@
 from axlearn.common.utils import Nested, Tensor
 
 
+def _is_valid_ctc_seq(paddings: Tensor, target_labels: Tensor, target_paddings: Tensor) -> Tensor:
+    """Returns for per example sequence if it passes validity check.
+
+    Note that `optax.ctc_loss` returns -logeps (default to 1e5) if the
+    input length is smaller than the label length plus number of
+    consectutive duplications, because we need an blank to token to transit
+    between the same tokens. When this condition is not met, it should be
+    considered as an invalid sequence and the loss should be ignored.
+
+    A validity check is passed if for an example when :
+      input.length >= labels.length + num(consecutive dup label tokens)
+
+    Args:
+      paddings:              [batch_size, num_frames], 0/1 Tensor, paddings of input frames.
+      target_labels:         [batch_size, num_frames], int32 Tensor.
+      target_paddings:       [batch_size, num_frames], 0/1 Tensor.
+
+    Returns:
+      A shape [b] float tensor indicating if each (input, label) pair is valid,
+      with a value of 1.0 indicating valid and 0.0 otherwise.
+    """
+    # [batch_size, ]
+    label_lengths = jnp.sum(1.0 - target_paddings, axis=-1)
+    # [batch_size, ]
+    input_lengths = jnp.sum(1.0 - paddings, axis=-1)
+    # [batch_size, num_frames - 1]
+    dups = (1.0 - target_paddings[:, 1:]) * (target_labels[:, :-1] == target_labels[:, 1:])
```

Comment:
I wonder whether we should consider `jnp.logical_or(target_paddings[:, :-1], target_paddings[:, 1:])` when dropping dups?

Reply:
As discussed offline, `dups` may be miscounted if target paddings appear on the leading duplicate labels. In normal cases, this shouldn't matter since paddings will be inferred from an OOV label, thus the paddings should cover both dups -- but in this case the API is taking a separate `target_paddings` Tensor which may not follow this assumption.

Reply:
Thanks. I have added a note in the docstring to explain the expectation of target_paddings,  and a TODO to support any paddings in the future.

Reply:
Sounds good, thanks.